`lie_bracket` and `adjoint_action`

mateuszbaran opened the pull request:
This PR adds two things that seem to be missing from our Lie group/Lie algebra support, as identified in #366 .

kellertuer commented:
We might need some more documentation, especially, it would be nice to have the general case ( https://en.wikipedia.org/wiki/Lie_bracket_of_vector_fields ) on Manifolds, too (at least have it in mind here).

mateuszbaran commented:
Yes, I'll add more documentation here. For this PR I'm going to restrict the Lie bracket to Lie algebras, and the more general case can be added later (I don't even know what could it be useful for).

sethaxen commented on src/groups/circle_group.jl:
```diff
@@ -14,6 +14,13 @@ invariant_metric_dispatch(::CircleGroup, ::ActionDirection) = Val(true)
 
+adjoint_action(::CircleGroup, p, X) = X
+
+function adjoint_action!(::CircleGroup, Y, p, X)
+    copyto!(Y, X)
+    return Y
+end
```
```suggestion
adjoint_action!(::CircleGroup, Y, p, X) = copyto!(Y, X)
```

sethaxen commented on src/groups/circle_group.jl:
```diff
@@ -46,6 +53,13 @@ function inverse_translate(
 
+lie_bracket(::CircleGroup, X, Y) = zero(X)
+
+function lie_bracket!(::CircleGroup, Z, X, Y)
+    fill(Z, 0)
+    return Z
+end
```
```suggestion
lie_bracket!(::CircleGroup, Z, X, Y) = fill!(Z, 0)
```

sethaxen commented on src/groups/group.jl:
```diff
@@ -484,6 +509,18 @@ function decorator_transparent_dispatch(
+"""
+    lie_bracket(G::AbstractGroupManifold, X, Y)
+
+Lie bracket between elements `X`, `Y` of the Lie algebra corresponding to the Lie group `G`.
```
```suggestion
Lie bracket between elements `X` and `Y` of the Lie algebra corresponding to the Lie group `G`.
```

sethaxen commented on src/groups/group.jl:
```diff
@@ -951,6 +988,13 @@ Base.:*(e::Identity{G}, p) where {G<:AdditionGroup} = e
 
+adjoint_action(::AdditionGroup, p, X) = X
+
+function adjoint_action!(::AdditionGroup, Y, p, X)
+    copyto!(Y, X)
+    return Y
+end
```
```suggestion
adjoint_action!(::AdditionGroup, Y, p, X) = copyto!(Y, X)
```

sethaxen commented on src/groups/group.jl:
```diff
@@ -988,6 +1032,13 @@ group_log(::AdditionGroup, q) = q
 
+lie_bracket(::AdditionGroup, X, Y) = zero(X)
+
+function lie_bracket!(::AdditionGroup, Z, X, Y)
+    fill(Z, 0)
+    return Z
+end
```
```suggestion
lie_bracket!(::AdditionGroup, Z, X, Y) = fill!(Z, 0)
```

sethaxen commented on src/groups/group.jl:
```diff
@@ -1015,6 +1066,13 @@ Base.:\(p, ::Identity{G}) where {G<:MultiplicationGroup} = inv(p)
 Base.:\(::Identity{G}, p) where {G<:MultiplicationGroup} = p
 Base.:\(e::E, ::E) where {G<:MultiplicationGroup,E<:Identity{G}} = e
 
+adjoint_action(G::MultiplicationGroup, p, X) = p * X / p
```
I think it's better to use the `translate_diff` and `inverse_translate_diff` machinery here and below, especially because e.g. for `SpecialOrthogonal` that is implemented more efficiently.

mateuszbaran replied:
Which machinery do you mean? I just checked that and I don't see anything that would make implementing this using `translate_diff` and `inverse_translate_diff` more efficient.

sethaxen replied:
What I mean is to define
```suggestion
function adjoint_action(G::MultiplicationGroup, p, Xₑ)
    e = make_identity(G, p)
    Xₚ = translate_diff(G, p, e, Xₑ, LeftAction())
    Y = inverse_translate_diff(G, p, p, Xₚ, RightAction())
    return Y
end
```
This is general enough actually to work for non-multiplication groups. But let's take the case of `SO(n)`, where all tangent vectors are represented as their translation to the Lie algebra. What Julia will execute is something like
```julia
function adjoint_action(G::SpecialOrthogonal, p, Xₑ)
    e = one(p)
    Xₚ = Xₑ
    Y = p * Xₚ * p'
    return Y
end
```
The `Xₚ * p'` is much more efficient than `Xₚ / p`, which is what I'm referring to.

If you defines this generally for `AbstractGroup`, then for `TranslationGroup`, this would simplify to
```julia
function adjoint_action(G::TranslationGroup, p, Xₑ)
    e = zero(p)
    Xₚ = Xₑ
    Y = Xₚ
    return Y
end
```
so you almost don't even need to special-case for `AdditionGroup`.

mateuszbaran replied:
Yes, that's a good point.

sethaxen commented on src/groups/group.jl:
```diff
@@ -1062,3 +1120,10 @@ function group_exp!(G::MultiplicationGroup, q, X)
+lie_bracket(::MultiplicationGroup, X, Y) = X * Y - Y * X
+
+function lie_bracket!(::MultiplicationGroup, Z, X, Y)
+    copyto!(Z, X * Y - Y * X)
```
```suggestion
    mul!(Z, X, Y)
    mul!(Z, Y, X, -1, true)
```

sethaxen commented on src/groups/group.jl:
```diff
@@ -1062,3 +1120,10 @@ function group_exp!(G::MultiplicationGroup, q, X)
 
 group_log!(::MultiplicationGroup, X::AbstractMatrix, q::AbstractMatrix) = log_safe!(X, q)
+
+lie_bracket(::MultiplicationGroup, X, Y) = X * Y - Y * X
```
```suggestion
lie_bracket(::MultiplicationGroup, X, Y) = mul!(X * Y, Y, X, -1, true)
```

sethaxen commented on src/groups/special_euclidean.jl:
`````diff
@@ -98,7 +98,14 @@ R & t \\
 ````
 
+This function embeds $\mathrm{SE}(n)$ in the general linear group $\mathrm{GL}(n+1)$.
+It is an isometric embedding and group homomorphism [^RicoMatrinex1988].
`````
```suggestion
It is an isometric embedding and group homomorphism [^RicoMartinez1988].
```

sethaxen commented on src/groups/special_euclidean.jl:
```diff
@@ -98,7 +98,14 @@ R & t \\
+
 See also [`screw_matrix`](@ref) for matrix representations of the Lie algebra.
+
+[^RicoMatrinex1988]:
```
```suggestion
[^RicoMartinez1988]:
```

sethaxen reviewed (commented):
Looks good! It may be the case that the approach for `SpecialEuclidean` works more generally for semidirect product groups. i.e. the normal subgroup's part of the lie bracket looks like the lie bracket of the normal subgroup. We'd have to think more carefully about the other part of the tangent vector though.

mateuszbaran commented:
> Looks good! It may be the case that the approach for `SpecialEuclidean` works more generally for semidirect product groups. i.e. the normal subgroup's part of the lie bracket looks like the lie bracket of the normal subgroup. We'd have to think more carefully about the other part of the tangent vector though.

There is such a thing as a semidirect sum of Lie algebras but it looks like in practice the structure can be most easily obtained by embedding is some higher dimensional general linear group. There are some more general details here: https://en.wikipedia.org/wiki/Lie_algebra_extension#By_semidirect_sum .

sethaxen commented:
> There is such a thing as a semidirect sum of Lie algebras but it looks like in practice the structure can be most easily obtained by embedding is some higher dimensional general linear group. There are some more general details here: https://en.wikipedia.org/wiki/Lie_algebra_extension#By_semidirect_sum .

Yeah that looks identical to what we see in the special case of semidirect products embedded in `GL(n+1)` with the affine representation, it just has the extra Lie bracket for the non-normal subgroup, which produces zero vectors in the case where we can use the affine representation. Seems like a good default to have.

mateuszbaran commented:
I've read a bit about the representation theory of the adjoint representation and it is very nontrivial. I am no longer sure this is the right design :confused: .

sethaxen commented:
> I've read a bit about the representation theory of the adjoint representation and it is very nontrivial. I am no longer sure this is the right design 😕 .

What in particular are your concerns?

mateuszbaran commented:
> What in particular are your concerns?

Mostly the fact that adjoint representations of many groups and algebras are not faithful, and that we still won't have any generic faithful representations that would work on `FVector`s (if that is even possible?).

mateuszbaran commented:
OK, after thinking a bit more about it I think these methods can be useful anyway. I'd still like to finish differentials of translations and adjoint action on SE(n) (primarily SE(3)).

mateuszbaran commented:
@sethaxen Where did you get the formula for `function translate_diff!(G::SemidirectProductGroup, Y, p, q, X, conX::LeftAction)` from? I'm trying to understand it and I'm actually surprised that we can't for some reason get `translate_diff` for SE(n) through embedding it in GL(n+1)? Your code gives a different translation part of the returned tangent vector.

sethaxen commented:
> @sethaxen Where did you get the formula for `function translate_diff!(G::SemidirectProductGroup, Y, p, q, X, conX::LeftAction)` from? I'm trying to understand it and I'm actually surprised that we can't for some reason get `translate_diff` for SE(n) through embedding it in GL(n+1)? Your code gives a different translation part of the returned tangent vector.

I don't remember or if I worked it out myself. Can you share a snippet that fails?

mateuszbaran commented:
I guess nearly any invocation of `translate_diff` would differ between your version and the embedding in GL(n+1) but here is what I'm currently using for testing:
```julia
julia> using Manifolds, LinearAlgebra

julia> G = SpecialEuclidean(3);

julia> t = Vector{Float64}.([1:3, 2:4, 4:6]);

julia> ω = [[1.0, 2.0, 3.0], [3.0, 2.0, 1.0], [1.0, 3.0, 2.0]];

julia> p = Matrix(I, 3, 3);

julia> Rn = Rotations(3);

julia> pts = [ProductRepr(ti, exp(Rn, p, hat(Rn, p, ωi))) for (ti, ωi) in zip(t, ω)];

julia> X = ProductRepr([-1.0, 2.0, 1.0], hat(Rn, p, [1.0, 0.5, -0.5]));

julia> q = ProductRepr([0.0, 0.0, 0.0], p);

julia> translate_diff(G, pts[1], q, X)
ProductRepr with 2 submanifold components:
 Component 1 =
  3-element Vector{Float64}:
   2.2112553975587987
   0.5176292379367051
   0.9178287088559302
 Component 2 =
  3×3 Matrix{Float64}:
    0.0  0.5   0.5
   -0.5  0.0  -1.0
   -0.5  1.0   0.0

julia> X
ProductRepr with 2 submanifold components:
 Component 1 =
  3-element Vector{Float64}:
   -1.0
    2.0
    1.0
 Component 2 =
  3×3 Matrix{Float64}:
    0.0  0.5   0.5
   -0.5  0.0  -1.0
   -0.5  1.0   0.0
```
When calculating things using the embedding in GL(n+1) the left `translate_diff` would leave `X` unchanged.

sethaxen commented:
I wonder if maybe the difference is the choice in representation. For `GL(n+1)`, all tangent vectors `X` at point `p` are represented instead as `p \ X`. And while `SO(n)` adopts the same convention `SE(n)` does not, representing instead the rotational part of the tangent vector as `R^T Ω` but doing nothing to the translational part.

mateuszbaran commented:
That might be it. Why did you choose that convention? I think having semidirect products easily identifiable as subgrups of GL(M) would make defining new ones much easier and consistent, and this convention makes this identification hard.

sethaxen commented:
> That might be it. Why did you choose that convention?

Because it involved less divisions by points for `LeftAction`, which would probably be the most commonly used one. Otherwise almost every operation with tangent vectors uses division by a point. For `SL(n)` it also makes checking the validity of a tangent vector easier/cheaper.

mateuszbaran commented:
I'm slowly working out the more general form of differential of the right translation but it's quite challenging. It seems like we need for that differential of the group action with respect to the group element, and even coming up with notation isn't easy. I'm trying to document things as I go:
![image](https://user-images.githubusercontent.com/2551062/122616099-0ac9e580-d08a-11eb-9101-1eac801d9f72.png)
but this formula at the bottom seems like a quite horrible way to write that differential :confused: .

mateuszbaran commented:
I've split out the "connection manifold" part of the metric manifold. Ref: https://github.com/JuliaManifolds/Manifolds.jl/issues/387 .

mateuszbaran commented:
I've finally estabished a homomorphism between SE(n) and GL(n+1) so the differential of the right translation is now verified (and actually derived) this way. I didn't manage to fully derive it generally for semidirect product groups.

mateuszbaran commented:
@dehann I've added an example of adjoint action on SE(3): https://github.com/JuliaManifolds/Manifolds.jl/pull/381/commits/a3b811b36da7f58769c6bd86095941ca15bc92c6 . It shows that you can go about it two different ways: classic formulas that work on `FVector`s of coefficients (like here: https://github.com/JuliaManifolds/Manifolds.jl/pull/381/files#diff-c856e9189bb4b28cd5d9c2645bc7e8f4ddd98605f91831781dbc1f8217e1b200R87-R95 ) or the general way which uses differentials of left and right translations (and is general for any Lie groups): https://github.com/JuliaManifolds/Manifolds.jl/pull/381/files#diff-10808c095fb9edee2db8898db210f688390ae05268469026dd5001b3e3016008R366-R394 and they both agree as tested here: https://github.com/JuliaManifolds/Manifolds.jl/pull/381/files#diff-691a7da5df480d8515c7004a1809fb2975f5a40ef62f860d679c312be0ecb306R205-R219

dehann commented:
Hi @mateuszbaran , oh wow -- will work through this thanks!  

The part i'm paying most attention to at the moment is how ProductGroup or SEmidirectProductGroup should be used in general, plus studying up on the Lie bracket.

mateuszbaran commented:
The testing failure seems to be due to Project.toml incompatibility between ColorTypes and VisualRegressionTests, I've open a PR that will hopefully fix this: https://github.com/JuliaPlots/VisualRegressionTests.jl/pull/32

mateuszbaran commented:
I've added custom transparency rules for `GroupManifold` but now I'm a bit unsure whether `GroupManifold` itself should conditionally override the connection or, rather, overriding the connection should be done through `ConnectionManifold`. Perhaps the latter would be more consistent?

mateuszbaran commented:
With the latest commit we now have the right parallel transport on `SO(n)` :tada: . Sadly, for more generic Cartan-Schouten connections, I've only found parallel transport along one-parameter subgroups.

mateuszbaran commented:
@kellertuer what do you suggest as a temporary workaround to the problem of VisualRegressionTests not being correctly resolved on Julia 1.4? I guess it may take a while before VisualRegressionTests work with the newest ColorTypes (an older version seems to be fine but Julia 1.4 doesn't seem to be able to downgrade it).

I can disable recipe testing, disable testing on Julia 1.4 or add ColorTypes 0.10 as an explicit dependency of Manifolds.jl.

 https://github.com/JuliaPlots/VisualRegressionTests.jl/pull/32

kellertuer commented:
I feel like temporarily using ColorTypes 0.10 (can we add a comment in the Project.toml to state that its temporary?) would be my favourite option.

mateuszbaran commented:
For some bizarre reason ColorTypes.jl sometimes causes trouble and sometimes it does not. Weird.

mateuszbaran commented:
I think this PR adds/changes everything I intended to do for now. It should be safe to release it as non-breaking.

kellertuer commented on src/manifolds/ConnectionManifold.jl:
```diff
@@ -0,0 +1,404 @@
+"""
+    gaussian_curvature(M::AbstractManifold, p, B::AbstractBasis; backend::AbstractDiffBackend = diff_backend())
+
+Compute the Gaussian curvature of the manifold `M` at the point `p` using basis `B`.
```
Can we maybe add a formula here?

mateuszbaran replied:
Done :slightly_smiling_face:

kellertuer commented on src/manifolds/ConnectionManifold.jl:
```diff
@@ -0,0 +1,404 @@
+    kwargs...,
+)
+
+function injectivity_radius(M::AbstractConnectionManifold, p)
```
Could this instead be done with transparency?

mateuszbaran replied:
Actually I don't think this is even correct, I just copied it from the metric manifold. I'd prefer to resolve this in a separate PR.

kellertuer replied:
Ok, that's also fine with me.

kellertuer commented on src/manifolds/ConnectionManifold.jl:
```diff
@@ -0,0 +1,404 @@
+    ricci_curvature(M::AbstractManifold, p, B::AbstractBasis; backend::AbstractDiffBackend = diff_backend())
+
+Compute the Ricci scalar curvature of the manifold `M` at the point `p` using basis `B`.
+"""
```
Can we add a formula here or a reference to the tensor?

kellertuer commented on src/manifolds/ConnectionManifold.jl:
```diff
@@ -0,0 +1,404 @@
+
+Compute the Riemann tensor ``R^l_{ijk}``, also known as the Riemann curvature
+tensor, at the point `p`. The dimensions of the resulting multi-dimensional
+array are ordered ``(l,i,j,k)``.
```
Could we add a formula here or a reference to wikipedia?

kellertuer reviewed (commented):
This turned out to be an impressive amount of modelling and work; great! And also great to see that we will have connections then.

I have a few remarks but most are documentation and only one on modelling.

mateuszbaran commented on docs/src/manifolds/metric.md:
````diff
@@ -17,6 +17,8 @@ Pages = ["metric.md"]
 Depth = 2
 ```
 
+Note that a metric manifold is an [`AbstractConnectionManifold`](@ref) with the [`LeviCivitaConnection`](@ref) of the metric $g$, and thus a large part of metric manifold's functionality relies on this.
````
Yes, I was positively surprised how easy it actually was to introduce connections.

mateuszbaran commented on src/groups/circle_group.jl:
```diff
@@ -14,6 +14,10 @@ invariant_metric_dispatch(::CircleGroup, ::ActionDirection) = Val(true)
 
 default_metric_dispatch(::MetricManifold{ℂ,CircleGroup,EuclideanMetric}) = Val(true)
 
+adjoint_action(::CircleGroup, p, X) = X
```
It should be because we may want to work on tangent vectors that are just numbers (and thus immutable).

mateuszbaran commented on src/groups/connections.jl:
```diff
@@ -0,0 +1,174 @@
+going through group identity are one-parameter subgroups. See [^Pennec2020] for details.
+
+[^Pennec2020]:
+    > X. Pennec and M. Lorenzi, “5 - Beyond Riemannian geometry: The affine connection
```
Yes, that's quite weird. I've taken this title from https://www.sciencedirect.com/science/article/pii/B9780128147252000121 IIRC and it has the chapter number in front of chapter title.

mateuszbaran commented on src/groups/connections.jl:
```diff
@@ -0,0 +1,174 @@
+"""
+abstract type AbstractCartanSchoutenConnection <: AbstractAffineConnection end
+
+"""
```
It's a bit complicated. Each Lie group has at least a one-parameter family of Cartan-Schouten connections, but if the Lie algebra is commutative then the are all the same. These plus, minus and zero connections belong to this family so we could technically have a struct with a number that represents all of these. In practice these three seem to be the most important so for performance I think they should have different types.

Moreover, some groups have Cartan-Schouten connections that don't belong to this one-parameter family but I have no idea what their significance is.

mateuszbaran replied:
There isn't much they have in common except that they all belong to the same one-parameter family, so the improving readability of the name would be the only point of this refactoring. The problem is, `CartanSchouten{<:Symbol}` might suggest that you can get any Cartan-Schouten connection using this type but actually it would only cover these three.

kellertuer replied:
Ah right, with the last argument, you convinced me. While I still feel the names are too long, the parameter approach would yield a wrong impression maybe.

mateuszbaran commented on src/groups/connections.jl:
```diff
@@ -0,0 +1,174 @@
+end
+
+"""
+    vector_transport_to(M::CartanSchoutenMinusGroup, ::Identity, X, q, ::ParallelTransport)
```
Good point. The formulas in [^Pennec2020] for vector transport in Cartan-Schouten connections are given along one-parameter subgroups but "minus" and "plus" connections give flat geometry so it should be safe to have them defined for `X` tangent at any point.

mateuszbaran commented on src/groups/connections.jl:
```diff
@@ -0,0 +1,174 @@
+end
+
+"""
+    vector_transport_direction(M::CartanSchoutenZeroGroup, ::Identity, X, d, ::ParallelTransport)
```
It could be slightly generalized but the issue is that [^Pennec] only describes parallel transport along one-parameter subgroups and the "zero" connection is in general not flat, so I don't know if there are explicit formulas for vector transport along arbitrary geodesics.

kellertuer replied:
I am also not sure, I was just surprised that is was documented for a different signature than it was implemented – in the other two cases, here it is ok.

mateuszbaran commented on src/groups/connections.jl:
```diff
@@ -0,0 +1,174 @@
+    > Analysis, X. Pennec, S. Sommer, and T. Fletcher, Eds. Academic Press, 2020, pp. 169–229.
+    > doi: 10.1016/B978-0-12-814725-2.00012-1.
+"""
+vector_transport_direction(
```
Very good question. In [^Pennec] there are formulas for `vector_tranport_to`, and for the plus and minus connections the only way I know to transport in a direction is taking the group logarithm of a point. On the other hand, the formula for the zero connection is quite a bit different -- you don't have to compute any logarithms if you have a direction.

Maybe there is a way to rewrite the formulas for the plus and minus connections to make it more similar, I didn't check.

kellertuer replied:
Ok, then let's leave it at that for now.

mateuszbaran commented on src/groups/connections.jl:
```diff
@@ -0,0 +1,174 @@
+    return translate_diff!(M.manifold, Y, dexp_half, p, Y, LeftAction())
+end
+
+function vector_transport_to!(
```
In general it's only correct along one parameter subgroups, so I restricted it to elements of the Lie algebra (even though it's a bit more general).

I didn't document it because, for consistency here, I've only put documentation for `vector_transport_direction` methods and this method of `vector_transport_to` is fairly straightforward. Should I just document this one method or explain the design differently?

mateuszbaran commented on src/groups/semidirect_product_group.jl:
```diff
@@ -112,6 +112,23 @@ function compose!(G::GT, x, e::E, ::E) where {GT<:SemidirectProductGroup,E<:Iden
     return identity!(G, x, e)
 end
 
+@doc raw"""
```
I'll change this, I just didn't pay enough attention here :slightly_smiling_face: .

mateuszbaran replied:
This is fixed now.

mateuszbaran commented on src/groups/special_euclidean.jl:
```diff
@@ -84,6 +84,15 @@ Base.@propagate_inbounds function _padvector!(
     return X
 end
 
+function adjoint_action(::SpecialEuclidean{3}, p, fX::TFVector{<:Any,VeeOrthogonalBasis{ℝ}})
```
I'll add documentation. This method is mostly about demonstrating possibilities of the interface, and only works on small input, so `SArray` would be the most likely argument type and then the non-mutating version can't be efficiently implemented through the mutating one.

mateuszbaran commented on src/groups/special_euclidean.jl:
```diff
@@ -386,3 +405,108 @@ function group_log!(G::SpecialEuclidean{3}, X, q)
     return X
 end
+
+function lie_bracket(G::SpecialEuclidean, X::ProductRepr, Y::ProductRepr)
```
I think we're going to need a lot more non-mutating variants explicitly implemented but maybe this one isn't actually particularly important.

mateuszbaran commented on src/groups/special_euclidean.jl:
```diff
@@ -386,3 +405,108 @@ function group_log!(G::SpecialEuclidean{3}, X, q)
+end
+
+@doc raw"""
+    SE_in_GL
```
Yes, I don't like this name but `SpecialEuclideanEmbeddedInGeneralLinear` sounds even worse.

kellertuer commented on src/manifolds/ConnectionManifold.jl:
```diff
@@ -190,6 +196,12 @@ end
 
 Compute the Ricci scalar curvature of the manifold `M` at the point `p` using basis `B`.
+The curvature is computed as the trace of the Ricci curvature tensor with respect to
+the metric, that is ``R=g^{ij}_R_{ij}`` where ``R`` is the scalar Ricci curvature at `p`,
```
What is `g^{ij}_R_{ij}` supposed to be? Currently this would yield a double subscripts error.

mateuszbaran replied:
Just a typo, I'm fixing it right now.

mateuszbaran replied:
It should be fine now.

mateuszbaran commented:
I think I've addressed all issues now, except the `SE_in_GL` which I don't know what to do about -- it looks like a great use case for `EmbeddedManifold` but there seems to be no nice name for it.

kellertuer commented:
What do we do about the vector transports that are documented for identity but implemented more general? I hope I have marked all resolved questions as resolved, so could you once more check the ones I did not yet mark?

For the name: What about `SpecialEuclideanInGeneralLinear`? Maybe a little too long but I thin nicer than the short form?

mateuszbaran commented:
I've improved documentation of vector transports today, is it still inconsistent somewhere? The currently unresolved comments are:

1. This: https://github.com/JuliaManifolds/Manifolds.jl/pull/381#discussion_r664315395 where the conclusion seems to be that the current design is OK.
2. Here: https://github.com/JuliaManifolds/Manifolds.jl/pull/381#discussion_r664316401, https://github.com/JuliaManifolds/Manifolds.jl/pull/381#discussion_r664316760, and this: https://github.com/JuliaManifolds/Manifolds.jl/pull/381#discussion_r664318000 which should be (I hope) resolved by changes I've pushed today.
3. Here: https://github.com/JuliaManifolds/Manifolds.jl/pull/381#discussion_r664575128, and I'll rename that constant to `SpecialEuclideanInGeneralLinear`.
4. https://github.com/JuliaManifolds/Manifolds.jl/pull/381#discussion_r664322782 and there: https://github.com/JuliaManifolds/Manifolds.jl/pull/381#discussion_r664326751 I've added references today, and I think it should be fine to have some explicit non-mutating variants for the special Euclidean, special orthogonal and translations groups for better interoperability with `SArray` and AD.

kellertuer commented on src/groups/connections.jl:
```diff
@@ -0,0 +1,182 @@
+    > Analysis, X. Pennec, S. Sommer, and T. Fletcher, Eds. Academic Press, 2020, pp. 169–229.
+    > doi: 10.1016/B978-0-12-814725-2.00012-1.
+"""
+vector_transport_to(M::CartanSchoutenPlusGroup, Y, ::Identity, X, q, ::ParallelTransport)
```
```suggestion
vector_transport_to(M::CartanSchoutenPlusGroup, ::Identity, X, q, ::ParallelTransport)
```

kellertuer commented on src/groups/connections.jl:
```diff
@@ -0,0 +1,174 @@
+end
+
+"""
+    vector_transport_to(M::CartanSchoutenPlusGroup, ::Identity, X, q, ::ParallelTransport)
```
Well this is – both in the documentation and the signature for the documentation, still done for `::Identity`, while the generic implementation is still done for a.. `p`, so my question remains whether this might be inconsistent?

mateuszbaran replied:
Right, it should be fixed now.

kellertuer commented:
Thanks for the summary, the ones where the line got changed too much did not show up in the “Files Changed” view so I must have missed them that way; I only have the one open one left I just commented on. Thanks for carefully working on all my remarks.

mateuszbaran commented:
Thanks for a review. This is a quite significant set of changes so it's good to have them checked :slightly_smiling_face:.

kellertuer commented:
Once the tests have passed, I think this can be merged, though we loose two lines in coverage (if you want, you could check for these) but it is within the limit of our tolerance.

mateuszbaran commented:
Great! These two lines reported as not covered look like false positives to me so I'd ignore that.

mateuszbaran commented:
Julia 1.4 CI seems broken :confused:

kellertuer commented:
It seems that the call to `ambigs = Test.detect_ambiguities(Base)` does not finish in the `our_base_ambiguities()` function (tested locally on my machine with Julia 1.4. Commenting that base ambiguities test out, the test here locally run fine on Julia 1.4

mateuszbaran commented:
Weird, from my tests it looks like `DoubleFloats` alone stops ambiguity detection from completion. I'll move this test to newer versions and bump the number of ambiguities accordingly.

mateuszbaran commented:
Well, on 1.6 with all the other packages loaded it also takes way too long.
